### PR TITLE
Fix last remaining MyPy errors

### LINKED ISSUE #19891: Re-enable MyPy
# Why Mypy re-enable

For a few weeks MyPy checks have been disabled after the switch to Python 3.7 (per https://github.com/apache/airflow/pull/19317).

We should, however, re-enable it back as it is very useful in catching a number of mistakes.

# How does it work

We 've re-added the mypy pre-commit now - with mypy bumped to 0.910. This version detects far more errors and we should fix them all before we switch the CI check back.

* mypy will be running for incremental changes in pre-commit, same as before. This will enable incremental fixes of the code changed by committers who use pre-commits locally

* mypy on CI runs in non-failing mode. When the main pre-commit check is run, mypy is disabled, but then it is run as a separate step (which does not fail but will show the result of running mypy on all our code). This will enable us to track the progress of fixes

# Can I help with the effort, you ask? 

We started concerted effort now and incrementally fix all the mypy incompatibilities - ideally package/by/package to avoid huge code reviews. We'd really appreciate a number of people to contribute, so that we can re-enable mypy back fully and quickly :).

# How can I help?

What you need is:

* checkout `main`
* `./breeeze build-image`
* `pip install pre-commit`
* `pre-commit install`

This will enable automated checks for when you do a regular contribution.  When you make your change, any MyPy issues will be reporteed and you need to fix them all to commit. You can also commit with `--no-verify` flag to skip that, bu, well, if you can improve airlfow a little - why not?  

# How can I help more ? 

You can add PRs that are fixing whole packages, without contributing features or bugfixes. Please refer to this issue #19891 and ideally comment below in the issue that you want to take care of a package (to avoid duplicate work).

An easy way to run MyPy check for package can be done either from the host:
```
find DIRECTORY -name "*.py" | xargs pre-commit run mypy --files
```

or from ./breeze shell:
```
mypy --namespace-packages DIRECTORY
```

# Current list of mypy PRs:

https://github.com/apache/airflow/pulls?q=is%3Aopen+is%3Apr+label%3Amypy

# Remaining packages

Here is the list of remaining packages to be "mypy compliant" generated with:

```
pre-commit run mypy --all-files 2>&1 | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g" | grep "error:" | sort | awk 'FS=":" { print $1 }' | xargs dirname | sort | uniq -c | xargs -n 2 printf "* [ ] (%4d) %s\n"
```

* [ ] (   1) airflow/api/common/experimental
* [ ] (   1) airflow/contrib/sensors
* [ ] (   1) airflow/example_dags
* [ ] (   1) airflow/jobs
* [ ] (   4) airflow/models
* [ ] (   1) airflow/providers/microsoft/winrm/hooks
* [ ] (   1) airflow/providers/ssh/hooks
* [ ] (   1) tests/providers/amazon/aws/hooks
* [ ] (   1) tests/providers/google/cloud/hooks

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by potiuk**
Closes: #19891

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
Seems like this one will also be a collaborative effort eventually, There are a few errors around "airflow.models.decorators" package which I am not entirely sure why they are there and whether I made the right fix.

The one remaining  - after I think I solved all other is 

```
airflow/models/dag.py:2118: error: Argument 1 to "partial" has incompatible
type "_TaskDecoratorFactory"; expected "Callable[..., Any]"  [arg-type]
    ...return cast("TaskDecoratorFactory", functools.partial(task, dag=self))
                                                             ^
airflow/models/dag.py:2118: note: "_TaskDecoratorFactory.__call__" has type "staticmethod"
Found 1 error in 1 file (checked 2704 source files)

ERROR: The previous step completed with error. Please take a look at output above 

```

But I am not sure if the fixes are added are good. Seems like `mypy` sometimes uses the stub and sometimes not and I have no idea why.

**Comment by potiuk** [comment]
Any help appreciated @uranusjr @ashb :D

File: airflow/decorators/__init__.py [hunk 1]
```diff
@@ -44,3 +45,7 @@ def __getattr__(self, name: str) -> TaskDecorator:
 
 
 task = _TaskDecoratorFactory()
+
+
+if TYPE_CHECKING:
+    TaskDecoratorFactory = _TaskDecoratorFactory
```
Comment (potiuk) [inline airflow/decorators/__init__.py]: This was added as part of resolving those errors: 

```
airflow/models/dag.py:91: error: Module "airflow.decorators" has no attribute
"TaskDecoratorFactory"; maybe "_TaskDecoratorFactory" or "TaskDecorator"? 
[attr-defined]
        from airflow.decorators import TaskDecoratorFactory
        ^
```

File: airflow/decorators/__init__.pyi [hunk 2]
```diff
@@ -22,11 +22,14 @@
 
 from typing import Any, Callable, Dict, Iterable, List, Mapping, Optional, TypeVar, Union, overload
 
-__all__ = ["task"]
+from airflow.models.dag import dag
```
Comment (potiuk) [inline airflow/decorators/__init__.pyi]: This was added to solve:

```
airflow/example_dags/tutorial_taskflow_api_etl.py:25: error: Module
"airflow.decorators" has no attribute "dag"  [attr-defined]
    from airflow.decorators import dag, task
    ^
```

**Comment by uranusjr** [comment]
The naming and structuring of `airflow/decorators/__init__.py` is a small disaster. Let’s leave it out of this PR, I can fix them separately and perhaps try to improve the situation a bit.

**Comment by potiuk** [comment]
Ehh. so you will be the last one :)

**Comment by potiuk** [comment]
@uranusjr  - remove it. Let's do it this way - since I also want to re-enable mypy and we should wait a few days with some "older PRs" to get "no-rebase" warning, I will continue rebasing that one until it gets "green". So feel free to unentangle the "decorators" disaster in the meantime

File: tests/www/api/experimental/test_dag_runs_endpoint.py [hunk 3]
```diff
@@ -97,12 +97,8 @@ def test_get_dag_runs_success_with_state_no_result(self):
         # Create DagRun
         trigger_dag(dag_id=dag_id, run_id='test_get_dag_runs_success')
 
-        response = self.app.get(url_template.format(dag_id))
-        assert 200 == response.status_code
-        data = json.loads(response.data.decode('utf-8'))
-
-        assert isinstance(data, list)
-        assert len(data) == 0
+        with pytest.raises(ValueError):
```
Comment (potiuk) [inline tests/www/api/experimental/test_dag_runs_endpoint.py:100]: I have a slight doubt about this one. This is "technically" breaking change with the experimental API, but one that should be very welcomed by any user. So I am tempted to leave it like that (I would treat it as a bugfix).

**Comment by potiuk** [comment]
All right @uranusjr . Decorators are all yours:

```
No need to rebuild the image: none of the important files changed

airflow/models/dag.py:91: error: Module "airflow.decorators" has no attribute
"TaskDecoratorFactory"; maybe "_TaskDecoratorFactory" or "TaskDecorator"? 
[attr-defined]
        from airflow.decorators import TaskDecoratorFactory
        ^
airflow/models/dag.py:2118: error: Argument 1 to "partial" has incompatible
type "_TaskDecoratorFactory"; expected "Callable[..., Any]"  [arg-type]
    ...return cast("TaskDecoratorFactory", functools.partial(task, dag=self))
                                                             ^
airflow/models/dag.py:2118: note: "_TaskDecoratorFactory.__call__" has type "staticmethod"
Found 2 errors in 1 file (checked 2704 source files)

ERROR: The previous step completed with error. Please take a look at output above 


No need to rebuild the image: none of the important files changed

airflow/example_dags/tutorial_taskflow_api_etl.py:25: error: Module
"airflow.decorators" has no attribute "dag"  [attr-defined]
    from airflow.decorators import dag, task
    ^
Found 1 error in 1 file (checked 403 source files)

ERROR: The previous step completed with error. Please take a look at output above 
```

**Comment by potiuk** [comment]
Right - rebased and fixed last two errors: @uranusjr @ashb - I fixed two small thing (not even errors but inconsistencies and MyPy complaints in the new `Operator` Interface - I am not sure if adding "owner" to interface is a good idea - I think think we do not need to add dag though (as get_dag() shoudl return it anyway):

The errors were:

```
airflow/models/base.py:67: error: "Operator" has no attribute "dag" 
[attr-defined]
                return self.dag.dag_id
                       ^
airflow/models/base.py:68: error: "Operator" has no attribute "owner" 
[attr-defined]
            return f"adhoc_{self.owner}"


```

File: tests/www/api/experimental/test_dag_runs_endpoint.py (hunk 3 above)
Comment (ashb) [reply tests/www/api/experimental/test_dag_runs_endpoint.py:100]: What bit throws the value error here? (It's not clear to from just looking at the code)

Oh, that `state=dummy` results in a ValueError -- wouldn't that result in a 500 returned, when instead it should be a 400?

File: airflow/models/dagrun.py [hunk 4]
```diff
@@ -511,8 +511,12 @@ def update_state(
             finished_tasks = info.finished_tasks
             unfinished_tasks = info.unfinished_tasks
 
-            none_depends_on_past = all(not t.task.depends_on_past for t in unfinished_tasks)
-            none_task_concurrency = all(t.task.max_active_tis_per_dag is None for t in unfinished_tasks)
+            none_depends_on_past = all(
+                not t.task.depends_on_past for t in unfinished_tasks  # type: ignore[has-type]
```
Comment (ashb) [inline airflow/models/dagrun.py:515]: I'm surprised this needs type ignoring.

Comment (ashb) [reply airflow/models/dagrun.py:515]: Oh, `t.task` could be None as far as typing is concenred?

**Review by ashb — APPROVED** [review]

**Comment by uranusjr** [comment]
Those two are easy, use `get_dag` instead.

**Pull request merged**